lib/ts/ArgParser.cc

```diff
@@ -0,0 +1,506 @@
+/** @file
+
+  Powerful and easy-to-use command line parsing for ATS
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "ArgParser.h"
+#include "ink_file.h"
+#include "I_Version.h"
+
+#include <iostream>
+
+namespace ts
+{
+ArgParser::ArgParser() {}
+
+ArgParser::ArgParser(std::string const &name, std::string const &description, std::string const &envvar, int arg_num,
+                     Function const &f)
+{
+  // initialize _top_level_command according to the provided message
+  _top_level_command = ArgParser::Command(name, description, envvar, arg_num, f);
+}
+
+ArgParser::~ArgParser() {}
+
+ArgParser::Command &
+ArgParser::top_command()
+{
+  return _top_level_command;
+}
+
+void
+ArgParser::add_global_usage(std::string const &usage)
+{
+  _global_usage = usage;
+}
+
+// a graceful way to output help message
+void
+ArgParser::help_message() const
```

Comment:
This should take the `argc` / `argv` as arguments so it can be called in context and not just at the top level. For that, it should be easy to pass in the original values.